Storage/StorageManager.php

```diff
@@ -225,6 +225,65 @@ public function storeFileFromLocalPath($type, $path, $newFilename = null, $unlin
         return $file;
     }
 
+    /**
+     * Takes an input stream and store it in permanent storage
+     *
+     * It is the callers responsibility to associate the returned entity with a relation and save it to the database
+     *
+     * @param integer $type
+     * @param string $stream
+     * @param string $originalFilename
+     * @param array $metadata
+     * @param string|null $newFilename
+     * @throws \RuntimeException If writing the file fails
+     * @return StorageFile
+     */
+    public function storeStream($type, $stream, $originalFilename, $metadata, $newFilename = null)
```

Comment:
After thinking about this for a while, I'm thinking that creating another `store*` function is further fragmenting our implementation here.  Basically, we'd be required to have a `storeThing` function for each type of Thing we're dealing with.  I think a better implementation would be to deprecate the two existing `storeUploadedFile` and `storeFileFromLocalPath` functions and introduce a new `store` function.

The store function should have the following signature:

```php
public function store($type, $content, array $options = []);
```

And should be added to the interface as well.  This function should then intelligently switch on the `$content` parameter and check for the following:

* Is is an `UploadedFile`? 
* Is it a local file on the file system?
* Is it a resource or StreamInterface (from PSR7)?
* Is it a string?

The `$content` parameter should have PHPDoc that includes all available types.  I think this would be able to handle a lot more future cases of sources for storing things.  Additionally, we can ensure we use streams internally to that `store` function as much as possible to avoid memory related issues.

The `$newFilename` and `$unlinkAfterStore` parameters should then be moved to the options array with intelligent defaults.  Additionally, overwriting things like metadata values, sizes, etc could also be passed in through the options array, otherwise, the store function would compute those itself.  The symfony OptionsResolver component can/should be used to make sure that the values provided through the options are of the correct type, and are known to the StorageManager.

Deprecating the `storeFileFromLocalPath` and `storeUploadedFile` functions should then call `store` internally.

Tests should still succeed as they do now, and we should add additional tests to ensure all of the `$file` metadata is set properly for all cases.

Reply:
I like this idea. It simplifies the interface while making it easy to extend in the future. There's an arbitrary amount where the number of different `storeXYZ()` functions becomes too much, let's just end it here and have one `store()` that works for all. With a clear php doc block it should be easy for the developer to understand what is allowed or not.

Keep the existing two functions for now but change them to call `store()`, and in the next major version we can just drop them.